pkg/ccl/changefeedccl/checkpoint/checkpoint_test.go

```diff
@@ -164,6 +166,142 @@ func TestCheckpointMake(t *testing.T) {
 	}
 }
 
+func TestCheckpointRestore(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ts := func(wt int64) hlc.Timestamp {
+		return hlc.Timestamp{WallTime: wt}
+	}
+
+	for name, tc := range map[string]struct {
+		initialHighWater       hlc.Timestamp
+		oldCheckpointTs        hlc.Timestamp
+		oldCheckpointSpans     []roachpb.Span
+		newCheckpoint          jobspb.TimestampSpansMap
+		expectedError          error
+		expectedFrontierForOld checkpointSpans
+		expectedFrontierForNew checkpointSpans
+	}{
+		"old checkpoint ts is empty": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(0),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+			},
+			newCheckpoint: jobspb.TimestampSpansMap{
+				Entries: []jobspb.TimestampSpansMap_Entry{
+					{
+						Timestamp: ts(2),
+						Spans:     []roachpb.Span{{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+					},
+				},
+			},
+			expectedError: errors.New("checkpoint timestamp is empty"),
+		},
+		"new checkpoint ts is empty": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(2),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+			},
+			newCheckpoint: *jobspb.NewTimestampSpansMap(map[hlc.Timestamp]roachpb.Spans{
+				ts(2): {{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}},
+				ts(0): {{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+			}),
+			expectedError: errors.New("checkpoint timestamp is empty"),
+		},
+		"fine grained checkpoint has some spans with progress above the old checkpointed timestamp": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(1),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("e")},
+			},
+			newCheckpoint: *jobspb.NewTimestampSpansMap(map[hlc.Timestamp]roachpb.Spans{
+				ts(2): {{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+					{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+				ts(1): {{Key: roachpb.Key("d"), EndKey: roachpb.Key("e")}},
+			}),
+			expectedFrontierForOld: checkpointSpans{
+				{span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("e")}, ts: ts(1)},
+			},
+			expectedFrontierForNew: checkpointSpans{
+				{span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}, ts: ts(2)},
+				{span: roachpb.Span{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}, ts: ts(2)},
+				{span: roachpb.Span{Key: roachpb.Key("d"), EndKey: roachpb.Key("e")}, ts: ts(1)},
+			},
+			expectedError: nil,
+		},
+		"same as above but with non-zero initial highwater": {
```

Comment:
nit: repeat yourself instead of saying "same as above" so that when the test case fails you can tell what it is from just the name

Reply:
Done.